src/thermo/PengRobinson.cpp

```diff
@@ -0,0 +1,834 @@
+    for (size_t k = 0; k < m_kk; k++) {
+        m_pp[k] = 0.0;
+        m_tmpV[k] = 0;
+        for (size_t i = 0; i < m_kk; i++) {
+            m_pp[k] += moleFractions_[i] * m_aAlpha_vec_Curr(k, i);
+            m_tmpV[k] += moleFractions_[i] * m_aAlpha_vec_Curr(k, i) *(m_dalphadT_vec_Curr[i] / m_alpha_vec_Curr[i]);
+        }
+        m_pp[k] = m_pp[k] * m_dalphadT_vec_Curr[k] / m_alpha_vec_Curr[k];
+    }
+
+
+    for (size_t k = 0; k < m_kk; k++) {
+        coeff1 = m_b_current * (m_pp[k] + m_tmpV[k]) - daAlphadT * m_b_vec_Curr[k];
+        sbar[k] += GasConstant * log(GasConstant * T / (refP * mv))
+                   + GasConstant
+                   + GasConstant * log(mv / vmb)
+                   + GasConstant * m_b_vec_Curr[k] / vmb
+                   - coeff1* log(vpb2 / vmb2) / den1
+                   - m_b_vec_Curr[k] * mv * daAlphadT / den2 / m_b_current;
+    }
+    calculatePressureDerivatives();
+    getPartialMolarVolumes(m_partialMolarVolumes.data());
+    for (size_t k = 0; k < m_kk; k++) {
+        sbar[k] -= m_partialMolarVolumes[k] * m_dpdT;
+    }
+}
+
+void PengRobinson::getPartialMolarIntEnergies(double* ubar) const
+{
+    // u_i = h_i - p*v_i
+    double* vbar = 0;
+    double p = pressure();
+    getPartialMolarEnthalpies(ubar);
+    getPartialMolarVolumes(vbar);
+    for (size_t k = 0; k < m_kk; k++) {
+        ubar[k] = ubar[k] - p*vbar[k];
+    }
+}
+
+void PengRobinson::getPartialMolarCp(double* cpbar) const
+{
+    getCp_R(cpbar);
+    scale(cpbar, cpbar+m_kk, cpbar, GasConstant);
+}
+
+void PengRobinson::getPartialMolarVolumes(double* vbar) const
+{
+    for (size_t k = 0; k < m_kk; k++) {
+        m_pp[k] = 0.0;
+        for (size_t i = 0; i < m_kk; i++) {
+            m_pp[k] += moleFractions_[i] * m_aAlpha_vec_Curr(k, i);
+        }
+    }
+
+    double mv = molarVolume();
+    double vmb = mv - m_b_current;
+    double vpb = mv + m_b_current;
+    double fac = mv * mv + 2 * mv * m_b_current - m_b_current * m_b_current;
+    double fac2 = fac * fac;
+    double RTkelvin = RT(); 
+
+    for (size_t k = 0; k < m_kk; k++) {
+        double num = (RTkelvin + RTkelvin * m_b_current/ vmb + RTkelvin * m_b_vec_Curr[k] / vmb
+                      + RTkelvin * m_b_current * m_b_vec_Curr[k] /(vmb * vmb)
+                      - 2 * mv * m_pp[k] / fac
+                      + 2 * mv * vmb * m_aAlpha_current * m_b_vec_Curr[k] / fac2
+                     );
+        double denom = (pressure() + RTkelvin * m_b_current / (vmb * vmb)
+                        + m_aAlpha_current/fac
+                        - 2 * mv* vpb *m_aAlpha_current / fac2
+                       );
+        vbar[k] = num / denom;
+    }
+}
+
+double PengRobinson::speciesCritTemperature(double a, double b) const
+{
+    double tc;
+    if (b <= 0.0) {
+        tc = 1000000.;
+        return tc;
+    }
+    if (a <= 0.0) {
+        tc = 0.0;
+        return tc;
+    }
+    tc = a * omega_b / (b * omega_a * GasConstant);
+    return tc;
+}
```

Comment:
A more succinct way of writing this would be:
```suggestion
double PengRobinson::speciesCritTemperature(double a, double b) const
{
    if (b <= 0.0) {
        return 1000000.;
    } else if (a <= 0.0) {
        return 0.0;
    } else {
        return a * omega_b / (b * omega_a * GasConstant);
    }
}
```